activerecord/lib/active_record/errors.rb

```diff
@@ -123,6 +123,28 @@ class RecordNotUnique < WrappedDatabaseException
   class InvalidForeignKey < WrappedDatabaseException
   end
 
+  # Raised when a foreign key constraint cannot be added because the column type does not match the referenced column type.
+  class MismatchedForeignKey < WrappedDatabaseException
+    def initialize(message)
+      parts = message.scan(/`(\w+)`[ $)]/).flatten
+      if parts.size == 4
+        msg = "Column `#{parts[1]}` on table `#{parts[0]}` has a type of `#{column_type(parts[0], parts[1])}`. \n"
+        msg << "This does not match column `#{parts[3]}` on `#{parts[2]}`, which has type `#{column_type(parts[2].pluralize, parts[3])}`. \n"
+        msg << "To resolve this issue, change the type of the `#{parts[1]}` column on `#{parts[0]}` to be :integer. (For example `t.integer #{parts[1]}`). \n\n"
+      else
+        msg = "There is a mismatch between the foreign key and primary key column types.\n"
+        msg << "Verify that the foreign key column type and the primary key of the associated table match types.\n"
+      end
+      msg << message
+      super(msg)
+    end
+
+    private
+      def column_type(table, column)
+        ActiveRecord::Base.connection.columns(table).detect { |c| c.name == column }.sql_type
```

Comment:
My understanding of the flow is limited, but when developing, I couldn't find a way to access the connection when the error was thrown. If so, I imagine it's possible to pass it in.  Any ideas?  https://github.com/rails/rails/pull/26266/files#diff-98ce5dc0cd01f8b0435ff6feac10d8a8R752